### ISSUE TITLE
Versions don't compare properly

### ISSUE THREAD

**Issue by bmewing** [bmewing/requirements#31]
Need to determine how to compare semantic versions properly.
* 12 > 9
* 12a > 11ab
* 1c < 11

The current logic takes the existing version (e.g. 1.12-1a) and compares it against the required version (1.9) by splitting on . or -, appending characters until the two versions are the same length ([1,12,1a] vs [1,9,0]) and the uses the comparison operator on each component (1 vs 1, 12 vs 9, 1a vs 0). 

Because versions are originally stored as characters (and need to be since versioning doesn't restrict using letters), it's a character comparison which incorporates length which isn't relevant.

Anyway, need to make the version comparison robust against these issues.

**Comment by bmewing** [comment]
Another aspect of this issue is the logic of comparing full versions. So, if the requirement is > 0.9 and the currently installed version is 1.0, check_version would compute a [TRUE, FALSE] and return FALSE because it checks 'all' (when comparing on != it uses 'any' instead). 

So, with >/>=/</<= comparisons to the left are more important than comparisons to the right. But, when the requirement is > 0.9 and the available version is 0.9.1, it would compute [FALSE, FALSE, TRUE] (because 0 !> 0 and 9 !> 9). Despite this, 0.9.1 >> 0.9

So just need to figure out how to handle the logic properly.

**Comment by AdamSpannbauer** [comment]
One option here is to be pickier about what versions are valid.  R is pretty picky about version numbers.  See quote from [*Writing R Extensions*](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#The-DESCRIPTION-file) below.

> The mandatory ‘Version’ field gives the version of the package. This is a sequence of at least two (and usually three) non-negative integers separated by single ‘.’ or ‘-’ characters. The canonical form is as shown in the example, and a version such as ‘0.01’ or ‘0.01.0’ will be handled as if it were ‘0.1-0’. It is not a decimal number, so for example 0.9 < 0.75 since 9 < 75.

My first thought was "well that's just CRAN that'll be picky"; I tried to build a package locally with a version number breaking the rules, and `R CMD build` refuses to build.  So I don't think you can actually build an R package with some of the versions being tested.  Of course, it's probably possible to build it, but for our purposes, I think assuming these rules are being followed brings more value than a reduction in functionality.

I think it runs `base::numeric_version` on the `Version` section on the `DESCRIPTION` (or an equivalent process).  We can also use this function for the comparisons (see below).  Then our work is limited to adding in wild-card functionality.

```r
> '12.0.0' > '9.0' 
[1] FALSE
> numeric_version('12.0.0') > numeric_version('9.0.0') 
[1] TRUE
```

**Comment by bmewing** [comment]
This is good, I had assumed (without checking) that you could include letters in the versions. This will greatly simplify this issue.

**Comment by AdamSpannbauer** [comment]
This base R function looks pretty sexy for solving the wildcard equivalency checking.

```r
# Convert wildcard versions to regex
> glob2rx('1.*.3')
[1] "^1\\..*\\.3$"
```

**Comment by AdamSpannbauer** [comment]
@bmewing,

Concerning version comparisons.  Wildcards only need to be handled on one side of the expression, correct?

For example, in the expression `lhs > rhs`.  `lhs` will always come from `installed.packages` or `metacran` (which shouldn't have a `*`) and `rhs` will be user specified (possibly include `*`).  Checking equivalency becomes more complex if `*` is allowed on both sides.  Fairly simple process if only allowed on one.

**Comment by bmewing** [comment]
That's right, only the users specified requirement can contain a wildcard. Note too, `~=` is basically wildcard expansion (so `mgsub ~= 1` is the same as `mgsub == 1.*.*.*...`)

**Comment by AdamSpannbauer** [comment]
This seems most easily handled by using the `>=` equivalency [noted in the compatible release section of pep 440](https://www.python.org/dev/peps/pep-0440/#compatible-release)

![image](https://user-images.githubusercontent.com/16326083/55292353-b4dcca80-53b7-11e9-8a67-ce18e95fa586.png)

**Comment by AdamSpannbauer** [comment]
After re-reading relevant PEP 440, I think I misunderstood the usage of the comma when listing equivalency.  I read it as `>= 2.2, == 2.*` is the same as `>= 2.2 and == 2.*`, but I think it's intended to be the same as `>= 2.2 or == 2.*`.

Do you agree with the intepretation that `~= 2.2` is equivalent to `>= 2.2 or == 2.*`

**Comment by bmewing** [comment]
This brings up an interesting point - should we support multiple specifications for a package? Right now all the code only supports a single specification.

**Comment by bmewing** [comment]
> After re-reading relevant PEP 440, I think I misunderstood the usage of the comma when listing equivalency. I read it as `>= 2.2, == 2.*` is the same as `>= 2.2 and == 2.*`, but I think it's intended to be the same as `>= 2.2 or == 2.*`.
> 
> Do you agree with the intepretation that `~= 2.2` is equivalent to `>= 2.2 or == 2.*`

I think it's an AND, not an OR

3.0 >= 2.2 but 3.0 wouldn't be considered a compatible version with 2.*
2.3 >= 2.2 AND 2.3 == 2.* and you would expect 2.3 to be a compatible version with 2.*

**Comment by AdamSpannbauer** [comment]
makes a lot more sense for the inclusion of the operator in pep 440. thanks for clarifying

**Comment by AdamSpannbauer** [comment]
> This brings up an interesting point - should we support multiple specifications for a package? Right now all the code only supports a single specification.

I lean towards keeping it simple with single specification.  Not sure what PEP 440 rules are, but it seems like `~=` was created to avoid having to have multiple specifications.

I think if someones using something like `pkg >= 2.2, pkg <= 2.4` then their need is better suited by using the more restrictive `==`.

**Comment by AdamSpannbauer** [comment]
What was the intended difference between `compare_version` and `check_version`?  Based on the tests they look equivalent-ish but with the argument order flipped?

**Comment by bmewing** [comment]
`check_version` compared two full versions (e.g. `check_version('1.1.0', '1.0.1', '>=')`)
`compare_version` compared only one semantic element (e.g. `compare_version('1','1','>=')`)

This was because I wanted a function I could toss into a `vapply` to walk across the semantic elements of the version

**Comment by bmewing** [comment]
Fixed with #39
